text/3604-derive_c-enum_integer_conversions.md

```diff
@@ -0,0 +1,185 @@
+- Feature Name: derive_c-enum_integer_conversions
+- Start Date: 2024-04-01
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+I want to define that a C-Style Enum is an enum that is defined like this:
+```rust 
+enum CStyleEnum {
+  Variant1 = 10,
+  Variant2 = 20,
+  ...
+  VariantX = some_integer
+}
+```
+Since this is not documented by [the docs](https://doc.rust-lang.org/stable/std/keyword.enum.html) and [the book](https://doc.rust-lang.org/stable/book/ch06-01-defining-an-enum.html), but suported by the Rust Language, and is the topic of this RFC.
+
+The code snippet below is basically what I want to propose 
+```rust
+//derive easier conversions for C-Style Enum
+#[derive(TryFromInt, IntoInt)]
+enum CStyleEnum {
+  Variant1 = 10,
+  Variant2 = 20,
+  ...
+  VariantX = some_integer
+}
+
+assert_eq!(CStyleEnum::try_from(10), Ok(CStyleEnum::Variant1)); //this works out of the box
+assert_eq!((CStyleEnum::Variant1).into(), 10); // this works too
+```
+
+# Motivation
+[motivation]: #motivation
+
+**Why**:
+- reduce boilerplate code, e.g. writing manual TryFrom and Into impls for the Integers your C-Style Enum will be constructed from, and will be converted to via Into (For use in generic code, where you might not be able to use ``as`` easily)
+- Quality of life improvements, similar to [making deriving Default for enums possible](https://rust-lang.github.io/rfcs/3107-derive-default-enum.html)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Explain the proposal as if it was already included in the language and you were teaching it to another Rust programmer. That generally means:
+
+- No new named concepts
+- Explaining the feature 
+
+We have a C-Style enum we derive TryFromInt and IntoInt for (name is subject to change). This allows us to save time writing boilerplate for constructing our enum from integers, and converting our enum to integers
+```rust 
+#[derive(TryFromInt, IntoInt)]
+enum DNSOpCode {
+  //each variant is just a number, that we asociate with 
+  //one of these enum states/variants
+  StandardQuery = 0,
+  InverseQuery = 1
+  ServerStatus = 2,
+}
+```
+since the `DNSOpCode` enum maps each state to a number we can just use `as` to 
+turn it into a number, the statement below is equal to zero, because the `DNSOpCode::StandardQuery` enum variant maps to 0 (defined by our enum)
+```rust
+let opcode_num = DNSOpCode::StandardQuery as u8; //this returns 0
+```
+
+likewise, we can also use the Into trait to do the same thing
+opcode_num and opcode_num1 are the same, however opcode_num2 can be used in generic contexts where `as` cannot
```

Comment:
I don't see many APIs that consume `impl Into<u8>`. Is there an instance where such an API would be useful for a type like `DNSOpCode`?